### ISSUE TITLE
Question: Confusiong about DataEvent and TCPHandle.write

### ISSUE THREAD

**Issue by Merethiel** [skypjack/uvw#159]
In the DataEvent callback we receive const uvw::DataEvent& event which holds a unique_ptr to the data. So after the callback completes, is data internally managed and destroyed in the DataEvent object? Since I never allocated the data for reading it only seems natural it would be destroyed on its own as well?

Then in the TCPHandle.write it takes either a char* or a unique_ptr. In the documentation if I'm reading it correctly it states that if the buffer is a char* then

The request doesn't take the ownership of the data. Be sure that their
     * lifetime overcome the one of the request.

 and if it's a unique_ptr then 

The request takes the ownership of the data and it is in charge of delete
     * them.

so my question mainly had to do with if I want to create a simple echo server and I want to pass the data received from DataEvent into a TCPHandle.write, how exactly would I do it properly such that a unique_ptr is passed intot he write request and managed by the request so I do not have to delete it later.

My knowledge of C++ smart pointers and move semantics is vague but I did try this:
```
client->on<uvw::DataEvent>([](const uvw::DataEvent &event, uvw::TCPHandle &client){
            client.write(std::move(event.data.get()), event.length);
                    });
```
But for some reason I think that's wrong. Because .get() should return the raw pointer without giving up ownership so I'm actually just calling the write method which takes a `char*` and I have no idea if the lifespan of event.data.get() exceeds the call to write.
Is there a better way to do this? Or am I doing it correctly? Can I do something like the ConnectEvent example

```
client->on<uvw::DataEvent>([](const uvw::DataEvent &event, uvw::TCPHandle &client){
            auto dataWrite = std::unique_ptr<char[]>(new char[event.length]){event.data};
            client.write(std::move(dataWrite), event.length);
                    });
```

That example doesn't compile it's just how I imagined it to look.

**Comment by skypjack** [comment]
For an echo server, this should be the preferred approach:
```
client->on<uvw::DataEvent>([](const uvw::DataEvent &event, uvw::TCPHandle &client) {
            client.write(std::move(event.data), event.length);
});
```
That is, you should move the data you received and give the ownership to the write request. Otherwise, the data will go away after the callback returns with the risks of a dangling pointer for the write.

**Comment by Merethiel** [comment]
Thank you for the response!
That was the first thing I tried; however, I received this compiler error:
`Call to implicitly-deleted copy constructor of 'std::unique_ptr<char []>'` I realised if DataEvent is const reference, then `std::move(event.data)` will not compile. If I remove the const from the callback the above code compiles correctly. My main issue in the beginning was not realise the const reference. Thanks again!
